### PR TITLE
✨Add Saving Diagram Dialog

### LINKED ISSUE #362: UX allgemein: Alle Änderungen gehen bei Klick auf „Start Process“ verloren
**Beschreibe das Problem:**

Ein Endnutzer durchläuft die folgenden Aktionen: 

- er malt das „Hello World“ von Hand aus der Doku ab (weil es nicht mit BPMN Studio zusammen ausgeliefert wird)
- er startet das Diagramm durch Klick auf „Start Process“
- er stellt überrascht fest, dass seine Arbeit umsonst war
- es stellt sich heraus, er hätte vorher auf „Save“ drücken müssen

Die Tatsache, dass wir den Endnutzer hier derart in die Falle laufen lassen, schadet unserem Produkt und hinterlässt einen wenig professionellen Eindruck.

### LINKED ISSUE #363: UX allgemein: Alle Änderungen gehen bei Wechsel in Settings-Ansicht verloren
**Beschreibe das Problem:**

Ein Endnutzer durchläuft die folgenden Aktionen:

- er malt das „Hello World“ von Hand aus der Doku ab (weil es nicht mit BPMN Studio zusammen ausgeliefert wird)
- er klickt in der Statusbar auf „Settings“
- er stellt überrascht fest, dass seine Arbeit hierdurch verloren gegangen ist
- es stellt sich heraus, er hätte vorher auf „Save“ drücken müssen

Die Tatsache, dass wir den Endnutzer hier derart in die Falle laufen lassen, schadet unserem Produkt und hinterlässt einen wenig professionellen Eindruck.

### LINKED ISSUE #364: UX allgemein: Alle Änderungen gehen bei Wechsel in Plan-Ansicht verloren
**Beschreibe das Problem:**

Ein Endnutzer durchläuft die folgenden Aktionen:

- er malt das „Hello World“ von Hand aus der Doku ab (weil es nicht mit BPMN Studio zusammen ausgeliefert wird)
- er klickt in der Menubar auf "Plan"
- er kann nicht durch erneuten Klick auf "Design" zurück zu seinem Diagramm
- er stellt überrascht fest, dass seine Arbeit verloren gegangen ist
- es stellt sich heraus, er hätte vorher auf „Save“ drücken müssen

Die Tatsache, dass wir den Endnutzer hier derart in die Falle laufen lassen, schadet unserem Produkt und hinterlässt einen wenig professionellen Eindruck.

### LINKED ISSUE #365: UX allgemein: Alle Änderungen gehen bei Wechsel in TaskList-Ansicht verloren
**Beschreibe das Problem:**

Ein Endnutzer durchläuft die folgenden Aktionen:

- er malt das „Hello World“ von Hand aus der Doku ab (weil es nicht mit BPMN Studio zusammen ausgeliefert wird)
- er klickt in der Menubar auf "Dashboard" und dann auf "Task List"
- er kann nicht durch erneuten Klick auf "Design" zurück zu seinem Diagramm
- er stellt überrascht fest, dass seine Arbeit verloren gegangen ist
- es stellt sich heraus, er hätte vorher auf „Save“ drücken müssen

Die Tatsache, dass wir den Endnutzer hier derart in die Falle laufen lassen, schadet unserem Produkt und hinterlässt einen wenig professionellen Eindruck.

### LINKED ISSUE #366: UX allgemein: Alle Änderungen gehen bei Wechsel in ProcessInstanceList-Ansicht verloren
**Beschreibe das Problem:**

Ein Endnutzer durchläuft die folgenden Aktionen:

- er malt das „Hello World“ von Hand aus der Doku ab (weil es nicht mit BPMN Studio zusammen ausgeliefert wird)
- er klickt in der Menubar auf "Dashboard" und dann auf "Process Instance List"
- er kann nicht durch erneuten Klick auf "Design" zurück zu seinem Diagramm
- er stellt überrascht fest, dass seine Arbeit verloren gegangen ist
- es stellt sich heraus, er hätte vorher auf „Save“ drücken müssen

Die Tatsache, dass wir den Endnutzer hier derart in die Falle laufen lassen, schadet unserem Produkt und hinterlässt einen wenig professionellen Eindruck.

### LINKED ISSUE #478: 🐛Press to Run Process does not Save the current Process First
## How can others reproduce the issue?

1. Open a process,
2. edit something,
3. run it,
4. return to the design view,
5. see that your change is missing.

## Possible solution

Save before Running.

## What's your setup

Studio Versioin 3.0.0-b170.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by Vyperus**
## What did you change?

This branch adds a basic modal-dialog that asks the user if he/she wants to save the diagram if changes were made to the diagram.

Fixes #362 
Fixes #363 
Fixes #364 
Fixes #365 
Fixes #366 
Fixes #478

## Test it

- Open a diagram.
- Make a change to it, like moving a shape from a to b.
- Try to open another diagram via the process solution explorer.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.)
- [x] I've merged the `develop` branch into my branch before finishing this PR.
- [x] I've **not added any other changes** than the ones described above.
- [x] I've mentioned all **PRs, which relate to this one**

### CODE REVIEW

File: src/modules/processdef-detail/processdef-detail.html [hunk 1]
```diff
@@ -4,6 +4,21 @@
...
   <div class="bpmnio-container">
-    <bpmn-io view-model.ref="bpmn" xml.bind="process.xml"></bpmn-io>
+    <bpmn-io view-model.ref="bpmnio" xml.bind="process.xml"></bpmn-io>
+  </div>
+
+  <div class="modal" tabindex="-1" role="dialog" id="saveModal">
+    <div class="modal-dialog save-modal-dialog" role="document">
+      <div class="modal-content">
+        <div class="modal-header">
+          <h4 class="modal-title">Save Changes to diagram?</h4>
```
Comment (alexanderkasten) [inline src/modules/processdef-detail/processdef-detail.html]: I think it should be *changes

File: src/modules/processdef-detail/processdef-detail.scss [hunk 2]
```diff
@@ -46,3 +46,25 @@
...
+
+.show-modal {
+  display: block;
+}
+
+.custom-button {
+  margin: 10px 0;
+  background-color: white;
+  border: 1px solid darkgray;
+  padding: 5px 10px;
```
Comment (alexanderkasten) [inline src/modules/processdef-detail/processdef-detail.scss]: padding after margin and background-color to the bottom

File: src/modules/processdef-detail/processdef-detail.html (hunk 1 above)
Comment (cmg-dev) [reply src/modules/processdef-detail/processdef-detail.html]:
```html
<div class="modal-header">
  <h2 class="modal-title">Document Contains Changes.</h4>
</div>
<div class="modal-body">
  Save Changes to diagram?
</div>
<div class="modal-footer">
...
```

Bitte so machen.

Comment (cmg-dev) [reply src/modules/processdef-detail/processdef-detail.html]: ich würde immer versuchen alle Elemente des Modal sinnvoll zu nutzen.

File: src/modules/processdef-detail/processdef-detail.scss (hunk 2 above)
Comment (cmg-dev) [reply src/modules/processdef-detail/processdef-detail.scss]: @alexanderkasten Do we have a guide for this?

File: src/modules/processdef-detail/processdef-detail.ts [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
 import {bindable, computedFrom, inject} from 'aurelia-framework';
 import {Router} from 'aurelia-router';
 import {ValidateEvent, ValidationController} from 'aurelia-validation';
+
 import * as canvg from 'canvg-browser';
 import * as download from 'downloadjs';
+import * as $ from 'jquery';
 import * as beautify from 'xml-beautifier';
+
+import {IProcessDefEntity} from '@process-engine/process_engine_contracts';
```
Comment (cmg-dev) [inline src/modules/processdef-detail/processdef-detail.ts]: This is also a third-party reference ;)

But else: This is great!

Please move it to line 11.

File: src/modules/processdef-detail/processdef-detail.ts [hunk 4]
```diff
@@ -26,170 +28,179 @@ interface RouteParameters {
...
       }
     } catch (error) {
-      this.notificationService.showNotification(NotificationType.ERROR, `Error: ${error.message}`);
+      this._notificationService.showNotification(NotificationType.ERROR, `Error: ${error.message}`);
     }
   }
 
-  private validateXML(): void {
-    const registry: Array<IShape> = this.bpmn.modeler.get('elementRegistry');
+  private _validateXML(): void {
```
Comment (cmg-dev) [inline src/modules/processdef-detail/processdef-detail.ts]: I want this documented. How does this validation work?

You can add an issue and todo, and do the work later.

File: src/modules/processdef-detail/processdef-detail.ts [hunk 5]
```diff
@@ -211,28 +222,28 @@ export class ProcessDefDetail {
     });
   }
 
-  public async exportBPMN(): Promise<void> {
-    const xml: string = await this.bpmn.getXML();
+  private async _exportBPMN(): Promise<void> {
```
Comment (cmg-dev) [inline src/modules/processdef-detail/processdef-detail.ts]: Documentation see above.

Comment (robpal94) [reply src/modules/processdef-detail/processdef-detail.ts]: When reading the prototype of the function, its not really clear what this function does. Does it export a diagram as a xml? a jpg? a png? 

A neat thing about documentation comment using the `/** ... */` syntax is, that many text editors can display this docstring when pressing `ctrl + space` so you dont have to navigate to this method to look what it does ;)

Comment (Vyperus) [reply src/modules/processdef-detail/processdef-detail.ts]: As the methods states it exports a bpmn-file.

File: src/modules/processdef-detail/processdef-detail.html [hunk 6]
```diff
@@ -4,6 +4,21 @@
...
+        </div>
+        <div class="modal-footer">
+          <button type="button" class="custom-button" data-dismiss="modal" id="dontSaveButton">Don't Save</button>
+          <button type="button" class="custom-button" data-dismiss="modal" id="saveButton">Save</button>
+          <button type="button" class="custom-button" data-dismiss="modal" id="cancelButton">Cancel</button>
+        </div>
+      </div>
+    </div>
   </div>
```
[the remark this answers is not in the document]
Comment (Vyperus) [reply src/modules/processdef-detail/processdef-detail.html]: done

File: src/modules/bpmn-io/bpmn-io.scss [hunk 7]
```diff
@@ -39,7 +39,7 @@
 
 .resize-div {
   position: absolute;
-  height: 100%;
+  height: 90%;
```
Comment (SteffenKn) [inline src/modules/bpmn-io/bpmn-io.scss]: Why did you set the height of the resize div to 90%?
This makes resizing the property panel at the bottom impossible.

Comment (Vyperus) [reply src/modules/bpmn-io/bpmn-io.scss]: This was a quick fix for the scrolling of bpmn-studio, i will remove it.

**Comment by Vyperus** [comment]
The bug mentioned by @SteffenKn is actually a reported bug in the aurelia router: http://localhost:9000/#/processdef/eb3c8ae8-ba3a-428a-88a3-48316eadf822/detail

**Comment by SteffenKn** [comment]
I think this is not the same bug as I am talking about. The Bug you mentioned is about navigating back twice.
The bug I mentioned is that navigating back navigates to the same page when you canceled the confirmation of a call activity before.
The bug you mentioned only happens when you cancel the back navigation.

**Pull request merged**